src/MessagePack.UnityClient/Assets/Scripts/MessagePack/Formatters/CollectionFormatter.cs

```diff
@@ -535,6 +535,116 @@ protected override Queue<T> Complete(Queue<T> intermediateCollection)
+            // Count property is reading just private field _size;
+            var count = value.Count;
+            writer.WriteMapHeader(count);
+            if (count == 0)
+            {
+                return;
+            }
+
+            IFormatterResolver resolver = options.Resolver;
+            IMessagePackFormatter<TElement> elementFormatter = resolver.GetFormatterWithVerify<TElement>();
+            IMessagePackFormatter<TPriority> priorityFormatter = resolver.GetFormatterWithVerify<TPriority>();
+            PriorityQueue<TElement, TPriority>.UnorderedItemsCollection items = value.UnorderedItems;
+            using PriorityQueue<TElement, TPriority>.UnorderedItemsCollection.Enumerator e = items.GetEnumerator();
+            while (e.MoveNext())
+            {
+                writer.CancellationToken.ThrowIfCancellationRequested();
+                var pair = e.Current;
+                elementFormatter.Serialize(ref writer, pair.Element, options);
+                priorityFormatter.Serialize(ref writer, pair.Priority, options);
+            }
+        }
+
+        public PriorityQueue<TElement, TPriority>? Deserialize(ref MessagePackReader reader, MessagePackSerializerOptions options)
+        {
+            if (reader.TryReadNil())
+            {
+                return default;
+            }
+
+            var count = reader.ReadMapHeader();
+            if (count == 0)
+            {
+                return new PriorityQueue<TElement, TPriority>(comparer);
+            }
+
+            IFormatterResolver resolver = options.Resolver;
+            IMessagePackFormatter<TElement> elementFormatter = resolver.GetFormatterWithVerify<TElement>();
+            IMessagePackFormatter<TPriority> priorityFormatter = resolver.GetFormatterWithVerify<TPriority>();
+
+            // https://github.com/dotnet/runtime/blob/a6bf4e4b94db9c1e28f50393e0a1943892e34bed/src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs#L160C22-L160C47
+            // EnumerableHelpers.ToArray is called for IEnumerable<ValueTuple<TElement, TPriority>>.
+            var sharedBuffer = ArrayPool<ValueTuple<TElement, TPriority>>.Shared.Rent(count);
+            try
+            {
+                options.Security.DepthStep(ref reader);
+                try
+                {
+                    if (sharedBuffer.Length == count)
+                    {
+                        for (var i = 0; i < sharedBuffer.Length; i++)
+                        {
+                            reader.CancellationToken.ThrowIfCancellationRequested();
+                            sharedBuffer[i].Item1 = elementFormatter.Deserialize(ref reader, options);
+                            sharedBuffer[i].Item2 = priorityFormatter.Deserialize(ref reader, options);
+                        }
+
+                        return new PriorityQueue<TElement, TPriority>(sharedBuffer, comparer);
+                    }
+                    else
```

Comment:
I don't see the value in having this written twice with the only difference being skipping of creating the `ArraySegment`. You could initialize the array with just one bit of code, and then you could decide whether to pass the original array in, or create a span for a subset of it, based on the length at the tail end.
I gather that the reason you bother at all is to avoid boxing the `ArraySegment` unnecessarily, but this can still be achieved without duplicating anything except the actual call to the queue constructor.

Reply:
You are absolutely right.